### PR TITLE
Update for Gnome Shell 46

### PR DESCRIPTION
**Pull request by sergio-costas**
add_actor() has been removed after being deprecated, and it must be replaced with add_child(). Also, there are two add methods that also have to be changed.

Tested in Gnome 46 alpha, in a Fedora RawHide system. I also tested it in Gnome 45, but doesn't work, so it must be only 46.

### CODE REVIEW

**Comment by sergio-costas** [comment]
In fact, using this patch in Gnome Shell 45 results in a tiny, empty dock, and a lot of messages in journal like this one:

    Can't update stage views actor <unnamed>[<XXXXXXX>:0xYYYYYYYY] is on because it needs an allocation.

**Comment by RyzenDew** [comment]
issues to report 

image 
With shrink the dash icons pop up over the dock line 

and without 
image
icons are too far up 

image
when transparent icons have weird circles

**Comment by sergio-costas** [comment]
@RyzenDew Are you using Gnome 45 or Gnome 46?

**Comment by sergio-costas** [comment]
Ok, there are some extra problems:

```
JS ERROR: TypeError: (intermediate value).cairo_set_source_color is not a function
_drawIndicator@file:///home/raster/.local/share/gnome-shell/extensions/ubuntu-dock@ubuntu.com/appIconIndicators.js:411:17
```
For some reason, it can't find that function. Also

```
Error: No signal 'actor-added' on object 'MetaWindowGroup'

Stack trace:
  _create@file:///home/raster/.local/share/gnome-shell/extensions/ubuntu-dock@ubuntu.com/utils.js:172:30


Error: No signal 'actor-removed' on object 'MetaWindowGroup'

Stack trace:
  _create@file:///home/raster/.local/share/gnome-shell/extensions/ubuntu-dock@ubuntu.com/utils.js:172:30
```

I'll try to fix them...

File: metadata.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
 "shell-version": [
-    "45"
+    "46"
```
Comment (3v1n0) [inline metadata.json:3]: We should still try to support 45 if the efforts are not big enough.

**Comment by sergio-costas** [comment]
More things to fix:

```
The GObject property St.ScrollView.hscroll is deprecated.
0 ensureActorVisibleInScrollView() ["file:///home/raster/.local/share/gnome-shell/extensions/ubuntu-dock@ubuntu.com/dash.js":1113:38]
```

**Comment by sergio-costas** [comment]
Fixed the cairo_set_source_color bug.

File: metadata.json (hunk 1 above)
Comment (sergio-costas) [reply metadata.json]: Fixed.

**Comment by sergio-costas** [comment]
And fixed the signal problem.

**Comment by sergio-costas** [comment]
@RyzenDew I'm unable to reproduce those errors. Can you try again with the new changes that I added, just in case?

**Comment by RyzenDew** [comment]
> @RyzenDew Are you using Gnome 45 or Gnome 46?

46 beta

**Comment by RyzenDew** [comment]
also no change with the newest change

**Comment by rastersoft** [comment]
@RyzenDew Can you tell me your dock configuration? (icon size...)

**Comment by RyzenDew** [comment]
this needs to be rebased on the beta cause the alpha is 1 month old and the beta has a few changes

**Comment by sergio-costas** [comment]
There are still several bugs in Gnome 46:
* The icons have incorrect size, as @ryzendev commented
* The "push to show" option doesn't work: if it is selected, pushing in the dock zone doesn't make it to pop up
* The overview mode is broken: this message is shown in the log: `clutter_actor_allocate: assertion '!isnan (real_allocation.x1) && !isnan (real_allocation.x2) && !isnan (real_allocation.y1) && !isnan (real_allocation.y2)' failed`

**Comment by sergio-costas** [comment]
The "push to show" option fails because the data sent from Gnome Shell is... odd... I don't know if it's a problem from Gnome Shell itself and its PressureBarriers, or is a problem due to being using VirtualBox. The problem is that, once the mouse cursor is at the screen border, it doesn't matter how hard you move it against the border, the distance indicated by the barrier is always 0.765625, so unless you achieve the required pressure in the first "push", the dock won't appear.

**Comment by sergio-costas** [comment]
I've been able to fix the Overview problem in Gnome 46. It was because `maybeAdjustBoxSize()` and `maybeAdjustBoxToDock()` were being called with `this.spacing` as the last parameter, but that property isn't defined anywhere, so the received value was `undefined`, and this broke the allocation values (it put NaN in several places).

I just added a check to set it to zero if the value is undefined, but maybe it is possible to just fully remove the parameter itself, because it seems to be used only when calling those two methods, and I can't find neither other places where those methods are called, nor other places where the property is set. @3v1n0 What do you think?

**Comment by sergio-costas** [comment]
I decided to convert this into a draft to avoid merges until it is fully finished.

**Comment by sergio-costas** [comment]
Ok, it is ready for reviewing.

File: appIconIndicators.js [hunk 2]
```diff
@@ -7,6 +7,12 @@ import {
     St,
 } from './dependencies/gi.js';
 
+if (!Clutter.cairo_set_source_color) {
+    Clutter.cairo_set_source_color = function(cr, sourceColor) {
+        cr.setSourceColor(sourceColor);
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (sergio-costas) [reply appIconIndicators.js:15]: Done.

File: theming.js [hunk 3]
```diff
@@ -353,13 +354,20 @@ class Transparency {
             this._base_actor_style = '';
 
 
+        if (GObject.signal_lookup("actor-added", global.window_group) == 0) {
+            var added_signal = "child-added";
+            var removed_signal = "child-removed";
+        } else {
+            var added_signal = "actor-added";
+            var removed_signal = "actor-removed";
```
[the remark this answers is not in the document]
Comment (sergio-costas) [reply theming.js:362]: The problem with const is that they scope is the block, so outside the if ... else ... won't exist. So I have to use let, defining them outside.

[the remark this answers is not in the document]
Comment (sergio-costas) [reply theming.js:362]: Done.

File: utils.js [hunk 4]
```diff
@@ -655,3 +655,9 @@ export function laterRemove(id) {
         Meta.later_remove(id);
 }
 
+export function addActor(obj, actor) {
+    if (obj.add_actor)
+        obj.add_actor(actor)
+    else
+        obj.add_child(actor)
+}
```
[the remark this answers is not in the document]
Comment (sergio-costas) [reply utils.js]: Done.

File: theming.js [hunk 5]
```diff
@@ -353,13 +354,20 @@ class Transparency {
             this._base_actor_style = '';
 
 
+        if (GObject.signal_lookup("actor-added", global.window_group) == 0) {
```
[the remark this answers is not in the document]
Comment (sergio-costas) [reply theming.js:357]: Done.

File: windowPreview.js [hunk 6]
```diff
@@ -386,10 +386,15 @@ class WindowPreviewMenuItem extends PopupMenu.PopupBaseMenuItem {
...
         });
-        box.add_child(overlayGroup);
-        box.add_child(labelBin);
+        if (box.add) {
+            box.add(overlayGroup);
+            box.add(labelBin);
+        } else {
+            box.add_child(overlayGroup);
+            box.add_child(labelBin);
+        }
```
[the remark this answers is not in the document]
Comment (sergio-costas) [reply windowPreview.js:395]: It isn't necessary, add_child() works. Changed.

File: docking.js [hunk 7]
```diff
@@ -51,6 +51,12 @@ export const State = Object.freeze({
     HIDING:  3,
 });
 
+if (!global.display.supports_extended_barriers) {
+    global.display.supports_extended_barriers = function() {
+        return true;
+    }
+}
```
[the remark this answers is not in the document]
Comment (sergio-costas) [reply docking.js]: This was an old patch. I changed it. But yes, it belongs to utils. It's used in two places, so I think that it's better to keep the function.

File: docking.js [hunk 8]
```diff
@@ -51,10 +51,11 @@ export const State = Object.freeze({
     HIDING:  3,
 });
 
-if (!global.display.supports_extended_barriers) {
-    global.display.supports_extended_barriers = function() {
-        return true;
-    }
+function _supports_extended_barriers() {
```
[the remark this answers is not in the document]
Comment (sergio-costas) [reply docking.js:54]: Done.

File: docking.js [hunk 9]
```diff
@@ -51,10 +51,11 @@ export const State = Object.freeze({
...
+    if (global.display.supports_extended_barriers)
+        return global.display.supports_extended_barriers();
+    const {capabilities} = global.backend;
+    return ((capabilities & Meta.BackendCapabilities.BARRIERS) !== 0);
```
[the remark this answers is not in the document]
Comment (sergio-costas) [reply docking.js:58]: Done.

File: docking.js [hunk 10]
```diff
@@ -2313,6 +2313,9 @@ export class DockManager {
             });
 
         const maybeAdjustBoxSize = (state, box, spacing) => {
+            // ensure that an undefined value will be converted into a valid one
+            if (!spacing)
+                spacing = 0;
```
[the remark this answers is not in the document]
Comment (sergio-costas) [reply docking.js:2318]: Done.

**Comment by sergio-costas** [comment]
> Great work but you missed the show apps icon it's still has black around it

Ops!!! Fixed.

File: _stylesheet.scss
[the remark this answers is not in the document]
Comment (sergio-costas) [reply _stylesheet.scss:1]: I'm not sure what do you mean... I really didn't change anything, just added the required elements to ensure that it works as expected...

[the remark this answers is not in the document]
Comment (sergio-costas) [reply _stylesheet.scss:1]: Oh, I see... _dock.scss is this...

[the remark this answers is not in the document]
Comment (sergio-costas) [reply _stylesheet.scss:1]: Done. Anyway, I see that there are some changes in the DashToDock scss file that aren't ported to the one in Yaru (specifically, setting `.dashtodock-app-well-preview-menu-item` differently for each orientation). Do you want me to create a patch for that?

Captura desde 2024-03-01 13-12-00

imagen

**Comment by 3v1n0** [comment]
Ah, for your info, I've landed this to ubuntu yesterday, so also consider the fixups I did at https://github.com/micheleg/dash-to-dock/tree/ubuntu-dock

**Comment by jbicha** [comment]
In case you weren't already aware, I'm missing the indicator next to running apps in the Ubuntu Dock. This is with 89ubuntu1 that is now in Ubuntu 24.04 LTS. And I'm using GNOME Shell 46~beta-0ubuntu2 and I'm still using the same Yaru theme as Ubuntu 23.10.

**Comment by rastersoft** [comment]
@jbicha I don't understand... Can you put me a screen capture?

**Comment by muhammad-ans** [comment]
There is also this weird square around app icons in overview when it is installed.
image

**Comment by jbicha** [comment]
Here's a a screenshot from Ubuntu 23.10 showing expected behavior. There is a circle showing that I have a single window of my file browser running and a circle showing that the Welcome to Ubuntu app is running. The file browser is highlighted to show that it is the active focused window. Neither the dots nor the highlight are showing in my Ubuntu 24.04 LTS.

image

**Comment by jbicha** [comment]
@muhammad-ans From your screenshot, it looks like you are using Fedora.

- What version of Fedora?
- What version of gnome-shell?
- Are you using the default Adwaita themes or something else?

**Comment by muhammad-ans** [comment]
> @muhammad-ans From your screenshot, it looks like you are using Fedora.
> 
>     * What version of Fedora?
> 
>     * What version of gnome-shell?
> 
>     * Are you using the default Adwaita themes or something else?

I am using fedora silverblue 40 (pre release).
Gnome shell version is gnome 46-beta 
Theme is default adwaita.

**Comment by rastersoft** [comment]
@jbicha Oh... I see... I did my tests in Fedora Rawhide, not in Ubuntu 24.04, and there it works fine. This monday I'll check it in an Ubuntu 24.04.

**Comment by sergio-costas** [comment]
@jbicha I installed it in a fresh 24.04 system, and it works nearly flawlessly... the only problem is the number of dots when there are several applications.

imagen

**Comment by sergio-costas** [comment]
@jbicha Ok, using the builtin theme, it works...

imagen

**Comment by sergio-costas** [comment]
@jbicha Are you using `make clean && make && make install` to install the extension?

**Comment by sergio-costas** [comment]
Mmm... but changing from "Default" to any other style, the dots do work...

imagen

**Comment by sergio-costas** [comment]
Ok, the "one dot" detail already happens in the official ubuntu-dock package in Gnome 45, so it's not a bug.

**Comment by jbicha** [comment]
@sergio-costas  What version of GNOME Shell are you using? I'm using GNOME Shell 46 Beta from my team's PPA. I'm using gnome-shell-extension-ubuntu-dock [89ubuntu1](https://launchpad.net/ubuntu/+source/gnome-shell-extension-ubuntu-dock/89ubuntu1)

**Comment by sergio-costas** [comment]
@jbicha Oh, ok... Although settings says "Gnome 46", the package is 45.3. I'll install the PPA.

**Comment by sergio-costas** [comment]
Ok, now I installed it and you are right: the theme isn't being applied...

**Comment by jbicha** [comment]
Yes, I can confirm that the Ubuntu Dock theming works much better for me if I use the GNOME session instead of Ubuntu. @3v1n0 Do you want me to open a Yaru issue for this? Are you able to do a Yaru theme update for Ubuntu 24.04 LTS this week?

**Comment by vanvugt** [comment]
Please include/merge 255ef10db9c330643188faca5ea82093f7ad0bb0

File: windowPreview.js [hunk 11]
```diff
@@ -359,16 +359,16 @@ class WindowPreviewMenuItem extends PopupMenu.PopupBaseMenuItem {
                 ? Clutter.ActorAlign.START : Clutter.ActorAlign.END,
             y_align: Clutter.ActorAlign.START,
         });
-        this.closeButton.add_actor(new St.Icon({icon_name: 'window-close-symbolic'}));
+        this.closeButton.add_child(new St.Icon({icon_name: 'window-close-symbolic'}));
```
Comment (vanvugt) [inline windowPreview.js:362]: Please change to `set_child` per 255ef10db9c330643188faca5ea82093f7ad0bb0

Comment (sergio-costas) [reply windowPreview.js]: Done. Thanks!

File: _stylesheet.scss [hunk 12]
```diff
@@ -247,11 +247,21 @@ $dock_style_modes: [null, shrink, extended, extended-shrink];
     }
 }
 
+.overview-tile {
+    margin: 0px;
+    padding: 0px;
+    border: 0px;
+}
+
```
Comment (3v1n0) [inline _stylesheet.scss:258]: How is this related to the app icons?

It would be worth a comment on the commit.

File: _stylesheet.scss
[the remark this answers is not in the document]
Comment (3v1n0) [reply _stylesheet.scss:1]: Oh, that's a miss apparently, thanks for noticing.

**Comment by 3v1n0** [comment]
Looks good, please before we can merge rebase the branch so that
 1. each commit passes `eslint` check
 2. We have no commits fixing a previous commit (so squash those doing that)
 3. Ideally we don't remove the 45 support and then add it back
 4. Still keep commits separated to help history

**Comment by sergio-costas** [comment]
@3v1n0 Check if the rebase is correct, please.

**Comment by vanvugt** [comment]
This needs rebasing again because eb25d700836dd247c01600df00ca5332c9a032e0 and below it have different commit IDs to that of master.

File: windowPreview.js [hunk 13]
```diff
@@ -359,7 +359,7 @@ class WindowPreviewMenuItem extends PopupMenu.PopupBaseMenuItem {
                 ? Clutter.ActorAlign.START : Clutter.ActorAlign.END,
             y_align: Clutter.ActorAlign.START,
         });
-        this.closeButton.add_child(new St.Icon({icon_name: 'window-close-symbolic'}));
+        this.closeButton.set_child(new St.Icon({icon_name: 'window-close-symbolic'}));
```
Comment (vanvugt) [inline windowPreview.js:362]: This commit should be squashed into the commit that it fixes, and a comment added explaining why `set_child` is needed (copy from 255ef10db9c330643188faca5ea82093f7ad0bb0).

**Comment by garrett** [comment]
Thanks for the extension and also for the work in progress of porting it to GNOME Shell 46!

I also see the app icon issue on basically stock Fedora Silverblue 40 beta, testing branch (`fedora:fedora/40/x86_64/testing/silverblue`, version `40.20240314.0 (2024-03-14T00:39:44Z)`, basecommit `b1c8abe190ad8f2d4e1df48d18898dc8003c76517e443ca5fd9381e3773fdcd7`).

I had assumed that this was a bug in GNOME and didn't even consider it could be this PR of dash to dock. Disabling dash to dock fixes the issue, so my completely wild guess is that it's probably something adjusting the style of the apps, but across all of shell instead of being restricted to just the dock app icons.

Without dash to dock:

image

With dash to dock, specifically this PR, in GNOME 46 RC:

image

It's more than just the app icons; it's also affecting the horizontal and vertical spacing too.

---

I also see the running dot issue, mentioned above, but I think it's a known issue already. Here's a screenshot (from Fedora, just for reference, as the above ones are in Ubuntu):

image

File: windowPreview.js (hunk 13 above)
Comment (sergio-costas) [reply windowPreview.js:362]: I did a cherry-pick with the original commit. But it isn't in master...

**Comment by sergio-costas** [comment]
@karuboniru @garrett It should be fixed now. Thanks for testing it!

**Comment by vanvugt** [comment]
If you're using >100% scale then there's also an upstream fix coming for the dots: https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/3245

**Comment by vanvugt** [comment]
The dots are still misplaced:

Screenshot from 2024-03-19 15-42-55

**Review by vanvugt — COMMENTED** [review]
Conditionally approved pending the dot-over-the-icon bug mentioned above. It still seems to be happening in GNOME 46.0 at least.

**Comment by sergio-costas** [comment]
@vanvugt But, if I get it right, the patch for the dot problem is for Gnome Shell itself...

**Comment by DScalpel** [comment]
Hi,
Thanks for the extension update. I have one bug to report.

The Customize the dash color seems to leave weird artifacts around each icon with brown color.

I am using Arch linux gnome unstable repo.

Screenshot from 2024-03-19 20-58-24

**Comment by sergio-costas** [comment]
@DScalpel I can't reproduce your bug... Did you update to the last version of THIS branch?

**Comment by bam80** [comment]
> Thanks for the extension update. 

It wasn't updated yet, right? I see no new releases.

**Comment by DScalpel** [comment]
> @DScalpel I can't reproduce your bug... Did you update to the last version of THIS branch?

hi, yes i pulled the latest changes before raising it as a bug.

**Comment by Calosis** [comment]
> The dots are still misplaced (in GNOME 46 only?):
> 
> Screenshot from 2024-03-19 15-42-55

Can confirm, I'm hitting this bug currently.

**Comment by vanvugt** [comment]
> @vanvugt But, if I get it right, the patch for the dot problem is for Gnome Shell itself...

That's a slightly different bug where the gap exists but is only half the correct size. Here however, the gap does not exist (or has negative spacing?)

**Comment by vanvugt** [comment]
Merged, but the dot bug on GNOME 46 is unresolved still. So it's not ready for release but at least we can all work more efficiently by focussing on the master branch.

**Comment by a1ex-ak** [comment]
Снимок экрана от 2024-03-20 12-18-35

**Comment by vanvugt** [comment]
This pull request is closed. Please subscribe to #2160 instead.

**Comment by stuarthayhurst** [comment]
> More things to fix:
> 
> ```
> The GObject property St.ScrollView.hscroll is deprecated.
> 0 ensureActorVisibleInScrollView() ["file:///home/raster/.local/share/gnome-shell/extensions/ubuntu-dock@ubuntu.com/dash.js":1113:38]
> ```

Did this get addressed? I'm still seeing it on my system, I'm happy to file a bug if that's not expected or tracked somewhere already :)

**Comment by vanvugt** [comment]
That's being tracked in https://github.com/micheleg/dash-to-dock/issues/2159

**Comment by a1ex-ak** [comment]
Tell me when the update for Gnome 46 will be released

**Comment by nechodom** [comment]
> Tell me when the update for Gnome 46 will be released

I already have Gnome 46. So I think you can download it.

**Comment by vanvugt** [comment]
dash-to-dock for GNOME 46 has not been released yet. Most likely because there are other issues we'd still like to fix, *and* we're working on other projects too.
